python_modules/dagster/dagster/_check/__init__.py

```diff
@@ -613,9 +613,22 @@ def opt_int_elem(
 # ##### INST
 # ########################
 
-# type-checking note: Attempting to use the passed type (Type[T] -> T) to infer the output type has
-# issues with abstract classes, so here we count on the incoming object to be typed before
-# this runtime validation check.
+# NOTE: inst() and opt_inst() perform narrowing, while inst_param() and opt_inst_param() do not. The
+# reason for this is that there is a trade-off between narrowing and passing type information
+# through untouched. The only working narrowing implementation will sometimes lose type information.
+# This is because not every static type can be expressed as a runtime-checkable type:
+#
+#     foo = Foo(Bar())  # type is Foo[Bar]
+#     inst(foo, Foo[Bar])  # illegal, can't pass parameterized types to inst()
+#     inst(foo, Foo)  # type is Foo; because we couldn't pass parameterized type, we lost info
+#
+# In contrast, we don't lose type information when we pass the type of the checked argument through
+# without modification.
+#
+# Because of this trade-off, it makes sense for inst() to perform narrowing but not inst_param().
+# With inst(), we rarely have rich type information at the callsite (if we did we wouldn't be
+# calling inst()). With inst_param(), on the other hand, we should always have rich type information
+# at the callsite from the type annotation, so we should never need to narrow.
```

Comment:
Should we just rename `inst` to `narrow`? I think it would make all this more intuitive,

Reply:
>Should we just rename inst to narrow? I think it would make all this more intuitive,

I don't think so, because (a) it would disrupt the existing naming scheme of "param" being a modifier to an existing method; (b) I think this will almost entirely "just work" as people want, so they won't need to dig into the details.

Reply:
I buy it 👍🏻